README.md
逛长沙通 - 长沙通手机App(非官方版)
==============
<img src="http://ufologist.github.io/cst/images/splash.jpg" width="270" height="480" /> 
<img src="http://ufologist.github.io/cst/images/forum-list.jpg" width="270" height="480" /> 
<img src="http://ufologist.github.io/cst/images/topic-list.jpg" width="270" height="480" /> 
<img src="http://ufologist.github.io/cst/images/topic.jpg" width="270" height="480" /> 

体验(Android版)
--------------
<a href="https://raw.github.com/ufologist/cst/master/release/cst.apk">
    <img src="http://ufologist.github.io/cst/images/download-QR-Code.png" />
</a>


简介
--------------
话说Tina非常喜欢刷[长沙通](http://www.cstong.net/), 每天就是盯着有什么帖子更新了啊.

但苦于只有网站, 没有手机App, 而手机上打开网站效果又不是很理想, 看起来有点不方便(白花花的流量啊).

有一天Tina抱怨说, 为什么长沙通不出手机App啊, 我想在手机上逛长沙通!

在网上搜索了一番后, 还真[没发现有长沙通手机App](http://cstong.net/read.php?tid=564208).

做为[新时代的手工艺人](http://www.ruanyifeng.com/blog/2008/01/the_future_of_web_startups_part_i.html),
那就动手做一个呗, 当是给Tina的玩具.

手艺不好, 请大家多多包涵:), 感谢[长沙通](http://www.cstong.net/)提供的高质量内容, 如有不妥请联系我.


版本
--------------
v0.1(Android版) 2013-12-21(开发周期2013-12-19 ~ 2013-12-21)

功能测试版, 定下UI和功能模块, 仅实现了最迫切的东西, 因此有所缺陷, 使用上也有些卡顿(HTML5开发, 没有优化-_-")

为了节约流量, 每个论坛板块只能看前10个帖子, 每个帖子只能看前11条跟帖(包含楼主), 还不能翻页啊...亲

只是希望尽早赶出一个可用的版本, 其他神马都是浮云了.


更多信息
--------------
* [项目主页](http://ufologist.github.io/cst/)
* [更新记录](https://github.com/ufologist/cst/blob/gh-pages/changelog.md)
* [问题记录](https://github.com/ufologist/cst/blob/gh-pages/ISSUE.md)

changelog.md
逛长沙通 - 长沙通手机App(非官方版) 更新记录
==============
v0.1(Android版) 2013-12-21(开发周期2013-12-19 ~ 2013-12-21)

功能测试版, 定下UI和功能模块, 仅实现了最迫切的东西, 因此有所缺陷, 使用上也有些卡顿(HTML5开发, 没有优化-_-")

为了节约流量, 每个论坛板块只能看前10个帖子, 每个帖子只能看前11条跟帖(包含楼主), 还不能翻页啊...亲

只是希望尽早赶出一个可用的版本, 其他神马都是浮云了.

assets/www/app.js
/**
 * app框架页面逻辑
 * 主要实现滑动效果
 * 
 * @author Sun
 * @version 0.1 2013-12-21
 */
(function() {
    var $slideWrapper = $('.slide-wrapper');
    var $innerWrapper = $('.inner-wrapper');
    var $forumList = $('.forum-list');
    var $topicList = $('.topic-list');
    var $topic = $('.topic');

    // 记录回退键按了几次, 连续(计算间隔时间)2次才确认退出
    var quitFlag = 0;

    function offsetSlide($el) {
        var left = 0;
        $el.each(function() {
            // TODO 得出的值不是百分比, 因此不能动态布局
            left += $(this).width();
        });
        $innerWrapper.css('left', -left);
    }

    function exitApp() {
        if (navigator.app) {
            quitFlag += 1;

            if (quitFlag == 1) { // 按一次回退只提示
                cordova.exec(null, null, 'ToastPlugin', 'alert', [{
                    "text": "(╥﹏╥) 这是要退出的节奏吗?",
                    "duration": "short"
                }]);
            } else { // 按两次以上退出app
                cordova.exec(null, null, 'ToastPlugin', 'alert', [{
                    "text": "╮(╯_╰)╭ 逛完就关太伤App自尊了",
                    "duration": "long"
                }]);
                navigator.app.exitApp();
            }

            // 延时重置记录
            setTimeout(function() {
                quitFlag = 0;
            }, 3000);
        }
    }
    
    // remove click delays on browsers with touch UIs
    new FastClick(document.body);

    document.addEventListener('deviceready', function() {
        document.addEventListener('backbutton', function(event) {
            var $prev = $('.slide.active').find('.prev');
            if ($prev.length > 0) {
                $prev.click();
            } else {
                exitApp();
            }
            event.preventDefault();
        }, false);
    }, false);

    // 滑动到下一页
    // 偏移外层包装使其定位到下一个滑动页
    // 需要偏移的距离 = 下一个滑动页之前所有滑动页宽度之和
    $slideWrapper.on('click', '.next', function(event) {
        var $prevSlides = $(this).parents('.slide').removeClass('active')
                                 .next('.slide').addClass('active')
                                 .prevAll('.slide');
        offsetSlide($prevSlides);
    });
    // 滑动到上一页
    // 需要偏移的距离 = 上一个滑动页之前所有滑动页宽度之和
    $slideWrapper.on('click', '.prev', function(event) {
        var $prevSlides = $(this).parents('.slide').removeClass('active')
                                 .prev('.slide').addClass('active')
                                 .prevAll('.slide');
        offsetSlide($prevSlides);
    });

    // 点击论坛介绍滑动到论坛帖子列表页
    $forumList.on('click', '.forum-intro', function(event) {
        $forumList.find('.forum-intro').removeClass('active');
        $(this).addClass('active');

        $topicList.find('.slide-hd .header').text($(this).find('.forumname').text());
        $forumList.find('.next').click();
    });

    // 初始化帖子列表的滚动条
    $topicList.find('.slide-bd').height($topicList.outerHeight() - $topicList.find('.slide-hd').outerHeight())
    new iScroll($topicList.find('.slide-bd')[0], {
        checkDOMChanges: true
    });

    // 初始化帖子列表的滚动条
    $topic.find('.slide-bd').height($topic.outerHeight() - $topic.find('.slide-hd').outerHeight())
    new iScroll($topic.find('.slide-bd')[0], {
        checkDOMChanges: true
    });
})();

src/com/urucas/plugins/ToastPlugin.java
package com.urucas.plugins;

/*!
 *  
 * ToastPlugin.java
 * Created By Urucas
 * 
 * The source Code is available under 
 * same licenses as javamail which are:
 * CDDL-1.0, GPL-2.0, BSD
 * 
 */

import org.apache.cordova.api.CallbackContext;
import org.apache.cordova.api.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.widget.Toast;

public class ToastPlugin extends CordovaPlugin {

        @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        
                if(action.equals("alert")) {
                        
                        JSONObject params = args.getJSONObject(0);
                        String text = params.getString("text");
                        String duration = params.getString("duration");
                        
                        this.alert(text, duration, callbackContext);
                return true;
        }
        return false;
    }
        
        private void alert(String text, String duration, CallbackContext callbackContext) {

                Context context = this.cordova.getActivity();
            
            int duration1;
            if(duration.toLowerCase().equals("long")) {
                    duration1 = Toast.LENGTH_LONG;
            }else{
                    duration1 = Toast.LENGTH_SHORT;
            }
            Toast toast = Toast.makeText(context, text, duration1);
            toast.show();
        }
}

src/com/github/ufologist/cst/activity/MainActivity.java
package com.github.ufologist.cst.activity;

import org.apache.cordova.Config;
import org.apache.cordova.DroidGap;

import android.annotation.TargetApi;
import android.os.Bundle;

import com.github.ufologist.cst.R;

public class MainActivity extends DroidGap {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        super.setIntegerProperty("splashscreen", R.drawable.splash);

        // XMLHttpRequest cannot load http://127.0.0.1:40582/[somerandomstring]. Origin null is not allowed by Access-Control-Allow-Origin. at null:1
        // http://stackoverflow.com/questions/12409440/phonegap-cordova-app-breaks-in-jelly-bean-access-control-allow-origin-and-seta
        super.init();
        if(android.os.Build.VERSION.SDK_INT > android.os.Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1) {
            fixJellyBeanIssues();
        }

        super.loadUrl(Config.getStartUrl(), 3000);
    }

    @TargetApi(16)
    protected void fixJellyBeanIssues() {
        System.out.println(super.appView.toString());
        try {
            super.appView.getSettings().setAllowUniversalAccessFromFileURLs(true);
        } catch(NullPointerException e) {
            System.out.println(e.toString());
        }
    }
}

assets/www/cst.js
/**
 * cst网站数据逻辑
 * 主要实现提取原始页面(如论坛帖子页面)中的数据重新组装到自己的页面上
 *
 * TODO 暂时以实现功能为目的, 优化什么都是浮云, 考虑引入RequireJS和Backbone
 * 
 * @author Sun
 * @version 0.1 2013-12-21
 */
(function() {
    var config = {
        tokenUrl: 'http://bbs.cstong.net/index.php?m=bbs&c=forumlist',
        tokenRegExp: /TOKEN\s*:\s*'(\w+)'/,
        extraToken: function(string) {
            return string.match(config.tokenRegExp)[1];
        },

        focusPageApi: 'http://bbs.cstong.net/index.php?m=cst&c=Api&a=loadFocusPage',
        pageSize: 20,

        forumlistUrl: 'http://bbs.cstong.net/index.php?m=bbs&c=forumlist'
    };

    /**
     * 获取token, 用于发送post请求
     */
    function getToken() {
        var dfd = $.Deferred();
        var promise = $.ajax({
            type: 'GET',
            url: config.tokenUrl,
            cache: false
        });
        promise.done(function(html) {
            dfd.resolve(config.extraToken(html));
        });
        return dfd.promise();
    }

    /**
     * 获取焦点数据(JSON数据接口)
     */
    function getFocusPage(topic, page) {
        return $.ajax({
            type: 'POST',
            url: config.focusPageApi,
            data: {
                tab: topic,
                page: page,
                pageSize: config.pageSize
            },
            dataType: 'json'
        });
    }

    // 获取聚焦数据
    // TODO 没实现UI
    // getToken().done(function(token) {
    //     $.ajaxSetup({
    //         data: {
    //             csrf_token: token
    //         }
    //     });
    //     getFocusPage('all', 1).done(function(data) {
    //         console.log(data);
    //         console.log(data.page, data.total, data.count);
    //     });
    // });

    /**
     * 将帖子图片替换成小图(节约流量), 可点击替换成大图
     */
    function clearPostImg($el, selector) {
        $el.find(selector).each(function() {
            var src = $(this).data().src;
            if (src.indexOf('attachment/thumb') != -1) {
                src = src.replace('attachment/thumb', 'attachment/thumb/mini');
            } else {
                src = src.replace('attachment', 'attachment/thumb/mini');
            }
            this.src = src;
        }).removeAttr('onclick onload style')
          .click(function() {
              this.src = $(this).data().src;
          });
    }

    /**
     * 组装活跃板块列表的HTML结构
     */
    function getForumIntroHtml($forumline) {
        var html = '';

        if ($forumline.length > 0) {
            var forumimage = $forumline.find('.forumimage').wrap('<div></div>').parent().html();
            var forumname = $forumline.find('.forumname').wrap('<div></div>').parent().html();

            html += '<div class="span4 intro-wrapper">';
            html +=     '<div class="forum-intro">';
            html +=         '<div>' + forumimage + '</div>';
            html +=         '<h2 class="forumname-hd">' + forumname + '</h2>';
            html +=     '</div>';
            html += '</div>';
        }

        return html;
    }

    /**
     * 请求论坛首页获取活跃板块的数据
     */
    function getForumlist() {
        var $forumListhd = $('.forum-list .slide-bd');

        $.fastScan(config.forumlistUrl).done(function($el) {
            var $forumitems = $el.find('.forumitem');
            var $html = $('<div></div>');

            for (var i = 0, length = $forumitems.length; i < length; i+=3) {
                var $forumitem1n = $forumitems.eq(i);
                var $forumitem2n = $forumitems.eq(i + 1);
                var $forumitem3n = $forumitems.eq(i + 2);

                var $row = $('<div class="row-fluid"></div>');
                $row.append(getForumIntroHtml($forumitem1n));
                $row.append(getForumIntroHtml($forumitem2n));
                $row.append(getForumIntroHtml($forumitem3n));

                $html.append($row);
            }

            $forumListhd.html($html.html());
            $forumListhd.find('img').each(function() {
                this.src = $(this).data().src;
            });
            $forumListhd.find('a').each(function() {
                this.href = $(this).data().src;
            });
            $forumList.find('.forum-intro').eq(0).click();
        });
    }

    var $forumList = $('.forum-list');
    $forumList.on('click', '.forumname a', function(event) {
        $(this).parents('.forum-intro').click();
        return false;
    });
    // 点击活跃板块滑动到板块帖子列表
    $forumList.on('click', '.forum-intro', function(event) {
        var forumUrl = $(this).find('.forumname a').attr('href');
        $('.topic-list .post-list').empty();
        $.fastScan(forumUrl).done(function($el) {
            $el.find('.thread_posts_list .reply, .thread_posts_list .lastreply').remove();
            // XXX 只提取了前10篇帖子
            $el.find('#J_posts_list tr:gt(9)').remove();
            $el.find('.threadimgole').parent().each(function() {
                $(this).css('height', 'auto');
            });
            $('.topic-list .post-list').html($el.find('#J_posts_list').html())
                                      .find('img').each(function() {
                                         this.src = $(this).data().src;
                                      })
                                      .end()
                                      .find('.title a').each(function() {
                                         this.href = $(this).data().src;
                                      });
        });
    });

    var $topicList = $('.topic-list');
    var $topic = $('.topic');
    $topicList.on('click', '.title a', function(event) {
        $(this).parents('tr').click();
        return false;
    });
    // 点击板块帖子滑动到帖子页面
    $topicList.on('click', 'tr', function(event) {
        $topicList.find('.next').click();
        $topic.find('.slide-hd .header').text($topicList.find('.slide-hd .header').text());

        var threadUrl = $(this).find('.title a').attr('href');
        $topic.find('.post').empty();
        $.fastScan(threadUrl).done(function($el) {
            // XXX 只提取了前11条跟帖(包含楼主)
            $el.find('.floor:gt(10)').remove();
            $topic.find('.post').html($el.find('.floor'))
                                .find('.J_avatar, .cstface').each(function() {
                                    this.src = $(this).data().src;
                                });
            clearPostImg($topic.find('.post'), '.J_post_img');
        });
    });

    getForumlist();
})();